pkg/bytecode/vm.go

```diff
@@ -63,11 +71,50 @@ func (vm *VM) Run() error {
 			globalIndex := ReadUint16(vm.instructions[ip+1:])
 			ip += 2
 			vm.globals[globalIndex] = vm.pop()
+		case OpAdd, OpSubtract:
+			if err := vm.executeBinaryOperation(op); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
+// executeBinaryOperation checks that both sides of the expression
+// are valid types and then processes the operation. An error is thrown
+// if the types are invalid for the operation.
+func (vm *VM) executeBinaryOperation(op Opcode) error {
+	right := vm.pop()
+	left := vm.pop()
+	rightType := right.Type()
+	leftType := left.Type()
+
+	// more cases to come (string, bool).
+	if leftType == parser.NUM_TYPE && rightType == parser.NUM_TYPE {
+		leftValue := left.(*numVal).V
+		rightValue := right.(*numVal).V
+		return vm.executeBinaryNumOperation(op, leftValue, rightValue)
+	}
+	return fmt.Errorf("%w %s %s", ErrUnsupportedType, leftType, rightType)
```

Comment:
hmm, this should never be let through by the parser - maybe even a panic is justified here if that makes your code easier? Up to you.

Reply:
If the parser changes the vm will not behave correctly and will panic in unusual ways that could be harder to debug than just reading a clear error message.

I discussed panic vs error in our implementation with josh and patrick yesterday and we feel that error reads better and there is not a material difference to the user (they just see an error message either way).

Reply:
ok. Could you please add a base error `ErrInternal` similar to the [evaluator package](https://github.com/evylang/evy/blob/1937ad097b260891e106c93f3d2a0f78610d68d8/pkg/evaluator/evaluator.go#L28) so that we separate internal errors more easily from other unavoidable user caused runtime errors?